data/transactions/transaction.go

```diff
@@ -343,12 +343,12 @@ func (tx Transaction) WellFormed(spec SpecialAddresses, proto config.ConsensusPa
 			return fmt.Errorf("tx.ForeignAssets too long, max number of foreign assets is %d", proto.MaxAppTxnForeignAssets)
 		}
 
-		if len(tx.ApprovalProgram) > proto.MaxAppProgramLen {
-			return fmt.Errorf("approval program too long. max len %d bytes", proto.MaxAppProgramLen)
+		if len(tx.ApprovalProgram) > ((1 + tx.ExtraProgramPages) * proto.MaxAppProgramLen) {
+			return fmt.Errorf("approval program too long. max len %d bytes", (1+tx.ExtraProgramPages)*proto.MaxAppProgramLen)
 		}
 
-		if len(tx.ClearStateProgram) > proto.MaxAppProgramLen {
-			return fmt.Errorf("clear state program too long. max len %d bytes", proto.MaxAppProgramLen)
+		if len(tx.ClearStateProgram) > ((1 + tx.ExtraProgramPages) * proto.MaxAppProgramLen) {
+			return fmt.Errorf("clear state program too long. max len %d bytes", (1+tx.ExtraProgramPages)*proto.MaxAppProgramLen)
```

Comment:
are we allowing a single page of `ExtraProgramPages` to be count towards both clear state and approval? One can pay for a page but utilize 2x space

Reply:
With these changes, paying extra will get you an extra 1k for each.  I intend to add changes that combine them later, and then each page will get to 2k total extra.